### PR TITLE
feat(routing/http): delegated IPNS server and client, IPIP 379

### PR DESCRIPTION
**Pull request by hacdias**
This implements the Boxo client and server for https://github.com/ipfs/specs/pull/379.

- [x] Requires #339 
- [x] Depends on fix: https://github.com/ipld/go-ipld-prime/pull/525
- [x] **Requires release** of https://github.com/ipld/go-ipld-prime (@rvagg)

### CODE REVIEW

**Comment by hacdias** [comment]
Note that this PR will have to be rebased if #339 is merged first.

File: routing/http/server/server.go
Comment (hacdias) [inline routing/http/server/server.go]: @lidel here I just named them `GetX` and `PutX`. I don't know how to feel about this considering the others are called `Find` and `Provide`. Perhaps, we should (in a separate PR), make the names consistent: `GetProviders`, `PutProviders`, etc. And I don't think there should be a specific `ProvideBitswap`. The implementer of `Provide` should be able to distinguish records and only use the ones they want.

File: routing/http/client/client.go
Comment (lidel) [inline routing/http/client/client.go]: :thought_balloon: I think its a good idea to land https://github.com/ipfs/boxo/pull/339 first and switch to `ipns.Name` here, removing direct dep. on libp2p

File: routing/http/client/client_test.go
Comment (lidel) [inline routing/http/client/client_test.go]:
```suggestion
func makeIPNSName(t *testing.T) (crypto.PrivKey, ipns.Name) {
```

File: routing/http/server/server.go
Comment (lidel) [reply routing/http/server/server.go]: - I like `Get` and `Put`, but consistency is more important. Maybe use `Find` and `Provide` for now, and cleanup everything later?
- +1 on  `ProvideBitswap` –  needs to be removed (or made protocol-agnostic): `boxo/routing/http/contentrouter/contentrouter.go#Provide*`  uses it blindly :see_no_evil:

**Review by lidel — CHANGES_REQUESTED** [review]

**Comment by hacdias** [comment]
I will address the feedback once we have the IPNS PRs complete then (or I might rebase it earlier if I have time).

**Comment by hacdias** [comment]
I rebased this on top of the IPNS PR to make the required changes. This also depends on a fix of `go-ipld-prime`, otherwise the mocking tests will fail due to an inconsistent behaviour between creating new integer nodes and decoding integer nodes: https://github.com/ipld/go-ipld-prime/pull/525.

Gateway sharness will fail until IPNS PR is merged.

File: routing/http/server/server.go
Comment (hacdias) [reply routing/http/server/server.go]: @lidel yeah, so `Provide` is already the generic version of `ProvideBitswap`. The problem with the naming `Provide` and `Find` is that now `Provide` should be renamed `ProvideProviders` to be consistent with `ProvideIPNSRecord`. 

Nevertheless, I think that cleanup can happen when we do the peer routing implementation. For now, the IPNS delegation is very nice because it's very contained.

File: examples/go.mod
Comment (lidel) [inline examples/go.mod]: @hacdias we are waiting for release as noted in https://github.com/ipld/go-ipld-prime/pull/525#issuecomment-1608818203, right?  (don't want to land unreleased commit in `main`, feels like asking for trouble, given breaking changes Rod  mentioned in linked comment)

File: routing/http/server/server_test.go
Comment (lidel) [inline routing/http/server/server_test.go]: While we are at this, can we add test for both V1+V2 and V2-only ([IPIP-428](https://github.com/ipfs/specs/pull/428))?
Will ensure server works with both and will save us headache in the future.

**Review by lidel — CHANGES_REQUESTED** [review]
Thanks for fixing tests, small nits below.

File: examples/go.mod
Comment (hacdias) [reply examples/go.mod]: @lidel yeah, that'd be great, but I have no idea if there's an ETA. cc @rvagg

File: routing/http/server/server_test.go
Comment (hacdias) [reply routing/http/server/server_test.go]: Updated both server and client tests.

File: examples/go.mod
Comment (rvagg) [reply examples/go.mod]: Yeah, sorry, this is a pressing concern for me too; I wanted to squeeze one more item into go-ipld-prime (https://github.com/ipld/go-ipld-prime/pull/530) before cutting a release since the releases there are a bit more involved than most other simpler libraries and we have a bit of a backlog of items already. I'm slowly getting back into some work this week but it probably won't be until next week to get this out; I hope this is OK.

**Comment by rvagg** [comment]
https://github.com/ipld/go-ipld-prime/pull/534

will get this out tomorrow unless something gets in the way

**Comment by hacdias** [comment]
@lidel we now have a new release of `go-ipld-prime` (thanks @rvagg) so I think this is ready for final review.

**Review by lidel — APPROVED** [review]
Thank you @hacdias, I think its good enough to merge and iterate further in https://github.com/ipfs/boxo/pull/422 +  https://github.com/ipfs/kubo/pull/9877.

ps. on windows [`TestClientOverMax` is flaky](https://github.com/ipfs/boxo/actions/runs/5931852593/job/16084571137#step:13:312), i've re-run it.

**Pull request merged**
